course5/homework06/simple_string.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+#include <iostream>
+#include <memory>
+
+namespace simple_string
+{
+
+class String
+{
+public:
+    String();
+    String(const char*);
+    String(const String& src);
+    String(String&& src);
+    String& operator=(const String& src);
+    String& operator=(String&& src);
+    String& operator=(const char * src);
+    String operator+(const char * src);
+    String operator+(const String& src);
+    String& operator+=(const char * src);
+    String& operator+=(const String& src);
+    bool operator==(const char * cmp) const;
+    bool operator==(const String& cmp) const;
+    bool operator!=(const char * cmp) const;
+    bool operator!=(const String& cmp) const;
+
+    char operator[](const int) const;
+    char& operator[](const int);
+
+    void Print() const;
+    int length() const;
+    const char *c_str() const;
+
+    typedef char * iterator;
+    typedef const char * const_iterator;
+    iterator begin() { return &buffer.get()[0]; }
```

Comment:
Knowing that for any pointer ```p``` and integer ```i```, the following expression are identical:
```p[i]``` === ```*(p + i)```
and
```&(*p)``` === ```p```
And knowing that ```buffer.get()``` is a pointer

How can you simplify this function?